andela/mervels-bn-backend
# #168131413 An Email should be sent to a user upon successful registration. User should be able to receive a verification link via email upon successful signup

#### What does this PR do?
- it send an email with verification link to the user
#### Description of Task to be completed?
- it sends a verification link to the user upon signup which is valid for one hour
- it sets the user as verified after clicking the link hence able to login
- it prevents anyone from signing in before verifying the email

#### How should this be manually tested?
- clone the repo
- checkout to the branch ft-email-verification-link-168131413
- run npm install to install the dependencies
- run npm test to run the tests
- run npm start:dev to start the development server
- signup with this payload to get a user who is currently not verified.
```
{	"firstName": "Robben",
	"lastName": "Bahati",
	"userEmail": "bahati.robben@andela.com",
	"userPassword": "Root1234"
}
```
- you should get something similar to this:
```
{
    "status": 201,
    "message": "Account has been created successfully",
    "data": {
        "firstName": "Robben",
        "lastName": "Bahati",
        "userEmail": "bahati.robben@andela.com",
        "userRoles": null,
        "userToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImJhaGF0aS5yb2JiZW5AYW5kZWxhLmNvbSIsImZpcnN0TmFtZSI6IlJvYmJlbiIsImxhc3ROYW1lIjoiQmFoYXRpIiwiaWF0IjoxNTY4MjI4ODI3fQ.qpqwCvtH51lxR2-v2pfJ-CsrztucNE3uzghuxAF5K20",
        "verification": {
            "message": "Verification link sent",
            "link": "http://localhost:PORT/api/v1/auth/verify/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImJhaGF0aS5yb2JiZW5AYW5kZWxhLmNvbSIsImZpcnN0TmFtZSI6IlJvYmJlbiIsImxhc3ROYW1lIjoiQmFoYXRpIiwiaWF0IjoxNTY4MjI4ODI3fQ.qpqwCvtH51lxR2-v2pfJ-CsrztucNE3uzghuxAF5K20"
        }
    }
}
```

- click the link sent into your email. You should get something like this
```
{
"status": 201,
"message": "Email verified succesfully",
"userEmail": "bahati.robben@andela.com"
}
```
- if you click it again you should get something like this:
```
{
"status": 409,
"error": "Email already verified"
}
```
- if the link is broken, or you delayed to verify(1 hr) you should get something like this:
```
{
"status": 401,
"message": "The link is invalid or has expired"
}
```
- Incase you didn't receive the link, just hit this enadpoint 'api/v1/auth/createLink'
```{
"email": "bahati.robben@andela.com"
}
```
- you should get something similar to this:
```
{
    "status": 200,
    "message": "email sent with verification link",
    "data": {
        "userEmail": "bahati.robben@andela.com",
        "link": "http://localhost:4500/api/v1/auth/verify/?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NTEsImVtYWlsIjoiYmFoYXRpLnJvYmJlbkBhbmRlbGEuY29tIiwiZmlyc3ROYW1lIjoiUm9iYmVuIiwibGFzdE5hbWUiOiJCYWhhdGkiLCJpYXQiOjE1NjgyMzE2MDN9.LPFDlTbi977Uz1WYQImUSuNzCLLh4cdnYr355xnduX4"
    }
}
```

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#168131413](https://www.pivotaltracker.com/story/show/168131413)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/39488016/64727042-40ff2c00-d4d8-11e9-8fb4-37d050adbdb1.png)
![image](https://user-images.githubusercontent.com/39488016/64727125-6db34380-d4d8-11e9-971b-3f0d9f662aac.png)
![image](https://user-images.githubusercontent.com/39488016/64728735-f41d5480-d4db-11e9-8248-140fcc69041f.png)
![image](https://user-images.githubusercontent.com/39488016/64728783-14e5aa00-d4dc-11e9-9921-1146e5566a14.png)
![image](https://user-images.githubusercontent.com/39488016/64728817-28911080-d4dc-11e9-9af4-0112b2de6e78.png)

#### Questions:
N/A
